---
title: Gimli
author: Ema
date: 11 December 2014
---

#Introduction


```gimli``` is a software which computes multiscale segmentations of methylation data.

The figure below corresponds to segmenting the same genomic region (chr10, from 1.1885e8,1.1893e8)
at three different levels of coarseness encoded by the parameter $\lambda$ : these three levels also broadly
correspond to three different inherent spatial scales for the segmentation.

```{r echo=F}
source("plot_segments.R")
```

#input file format

At the moment ```gimli``` processes chromosomes one at the time (i.e. each chromosome must be in a separate file).
The input format is very simple, for example

```
chr10 112000155 25 1
chr10 112000298 14 0
chr10 112000743 12 0
chr10 112000942 26 0
chr10 112001265 16 0
```

where the $4$ fields are chromosome, position along the chromosome, non converted reads, converted reads.

```gimli``` sits on the cluster at ```/home/devel/eraineri/shared/gimli```.

There's no need to create explicitly a file with the right format to compute the segmentation: one can use
the process substitution syntax of UNIX to prepare the input on the fly:

```
gimli <(zcat /scratch/devel/heath/Blueprint/Blueprint_8/G199_cpg.txt.gz  | awk '{if ($1=="chr10" && $2<=1.22e8 && $2>=1.12e8) print $1,$2,$6,$7}') > test.out 2>test.log
```

I am experimenting with a region on chr10 which goes from 1.12e8 to 1.22e8; this corresponds to $83821$ positions 
and is processed in less than $2s$.

#output file format

```gimli``` outputs the segments in a format similar to BED, with a lot of extra information. One example output line is:

```
chr10	116159696	116160685	8	0.9091	0.9654	1.0000	-3.05	-3.893	2.0000
```

which represents a segment on chromsome 10 from $116159696$ to $116160685$ containing $8$ CpGs,
with minimum $\theta$ equal to $0.9091$, maximum $\theta$ equal to $1$ and average (infact: maximum likelihood 
estimate of $\theta$) equal to $0.9654$.

Column $8$ ($-3.05$) is a confidence score on the segment ( this can go from $-\infty$ to $0$, the higher the more confident; but one can simply use the difference between maximum and minimum 
$\theta$ in practice) and the last field ($2$) is $\lambda$, the parameter controlling the coarseness of the segmentation ; typically
to larger $\lambda$s there correspond larger segments, see plot below.

```{r echo=F}
source("stats_segmentation.R")
```


#Examples

##PMD region.

For the above mentioned PMD region, this is the segmentation with $\lambda=1000$ across G199(green), G200(blue), G201(magenta),
G202(cyan).

![](G199.G200.G201.G202.png)





#to do

*on the biological side : check segments against imprinted regions, promoter regions, etc...

*on the statistical/computational side : 
	
	+introduce spatial constraints so as not to build long segments in regions where there are no CpGs.

	+extend the same approach to DMRs
