---
title: Gimli
author: Ema
date: 8 April 2015 
---

#Introduction


```gimli``` is a software which computes multiscale segmentations of methylation data.

Segmentations with very long segments will necessarily be more imprecise, while short segments
will more often correspond to better defined (less variable) methylation values.

The figure below corresponds to segmenting the same genomic region (chr10, from 1.1885e8,1.1893e8)
at three different levels of coarseness encoded by the parameter $\lambda$ : these three levels also broadly
correspond to three different inherent spatial scales for the segmentation.

```{r echo=F}
source("plot_segments.R")
```

#input file format

The input format is very simple, for example

```
chr10 112000155 25 1
chr10 112000298 14 0
chr10 112000743 12 0
chr10 112000942 26 0
chr10 112001265 16 0
```

where the $4$ fields are chromosome, position along the chromosome, non converted reads, converted reads.

```gimli``` sits on the cluster at ```/home/devel/eraineri/shared/gimli```.

The syntax to use ```gimli``` is, for example:

```gimli input.file 1:10:100:1000```

where ```input.file``` contains the methylation values and ```1:10:100:1000``` are $4$ values for $\lambda$
which will be tried one after the other.
Also, ```input.file``` can be ```-```if you want to read from the standard input.

There's no need to create in advance a file with the right format : one can use
the process substitution syntax of UNIX to prepare the input on the fly:

```
gimli <(zcat /scratch/devel/heath/Blueprint/Blueprint_8/G199_cpg.txt.gz  | awk '{if ($1=="chr10" && $2<=1.22e8 && $2>=1.12e8) print $1,$2,$6,$7}') 1:10:100:1000 > test.out 2>test.log
```
 
Dealing all the chromosomes of a typical CpG file ($\approx 20\times 10^6$ positions) 
takes around $15$ minutes if one specifies $4$ $\lambda$s.

#output file format

```gimli``` outputs the segments in a format similar to BED, with a lot of extra information. One example output line is:

```
chr1	10489	10525	4	0.7826	0.8667	0.9231	0.8669	0.004079	-1.238	-1.593	1.0000
```

which represents a segment on chromosome $1$ from $10489$ to $10525$ containing $4$ CpGs,
with minimum $\theta$ equal to $0.9091$, maximum $\theta$ equal to $1$,  maximum likelihood 
estimate of $\theta$ equal to $0.9654$, mean $\theta$ equal to $0.8669$ and variance of the
methylation values included in the segment corresponds to $0.004079$.
Column $10$ ($-1.238$) is a confidence score on the segment 
( this can go from $-\infty$ to $0$, the higher the more confident; but one can simply use the difference between maximum and minimum 
$\theta$ in practice) and the last field ($1.00$) is $\lambda$; as said, typically
to larger $\lambda$s there correspond larger segments, see plot below.

```{r echo=F}
source("stats_segmentation.R")
```


#Examples


1. segmentation of G199 and G202, chromosome 1.


![](G199.G202.chr1.gimli.png)

2. some DMRs (G199 vs G202) also on chr1, found by comparing segmentations and picking overlapping segments
which have different methylation levels.

![](G199.G202.100.200.dmr.png)

