#Introduction

GIML produces segmentations of methylation data generated by whole genome bisulfite sequencing
(WGBS).
The coarsness of the segmentation is controlled by a parameter called $\lambda$. The algorithm
behind GIML is similar to the one used in Vega (https://doi.org/10.1093/bioinformatics/btq586)
but it uses a different probabilistic model tailored for WGBS.

#Usage

giml <input-file> <lambdas>

input file is a file in tab-separated format:

chromosome position non-converted-reads converted-reads
if the file name is '-' giml will read from the standard input.

One can specifiy more tha one lambda by using ":" as a separator. The different
lambdas will be tried in sequence.

#output format

tab-separated format:

chromosome segment-start segment-end lambda index-start index-end methylation-mle

where index-start and index-end are indexes for the segment into a zero-based
array which stores the positions of the chromosome (useful for plotting and counting
the CpGs contained in a given segment).



